third_party/move/move-compiler/src/parser/syntax.rs

```diff
@@ -2749,14 +2794,70 @@ fn parse_struct_decl(
     })
 }
 
+/// Parse ability declarations:
+///
+///    Abilities = "has" <Ability> (, <Ability>)+
+fn parse_abilities(context: &mut Context) -> Result<Vec<Ability>, Box<Diagnostic>> {
+    if context.tokens.peek() == Tok::Identifier && context.tokens.content() == "has" {
+        context.tokens.advance()?;
+        parse_list(
+            context,
+            |context| match context.tokens.peek() {
+                Tok::Comma => {
+                    context.tokens.advance()?;
+                    Ok(true)
+                },
+                Tok::LBrace | Tok::Semicolon => Ok(false),
+                _ => Err(unexpected_token_error(
+                    context.tokens,
+                    &format!(
+                        "one of: '{}', '{}', or '{}'",
+                        Tok::Comma,
+                        Tok::LBrace,
+                        Tok::Semicolon
+                    ),
+                )),
+            },
+            parse_ability,
+        )
+    } else {
+        Ok(vec![])
+    }
+}
+
+/// Parse postfix ability declarations:
+///   PostfixAbilities = (<Abilities> ";")?
+fn parse_postfix_abilities(
+    context: &mut Context,
+    prefix_abilities: &mut Vec<Ability>,
+) -> Result<(), Box<Diagnostic>> {
+    let postfix_abilities = parse_abilities(context)?;
+    if !postfix_abilities.is_empty() {
+        consume_token(context.tokens, Tok::Semicolon)?;
+    }
+    if let (Some(sp!(_l1, _)), Some(sp!(l2, _))) =
+        (prefix_abilities.first(), postfix_abilities.first())
+    {
+        let msg =
+            "Conflicting ability declarations. Abilities must be declared either before or after \
+             the variant list, not both.";
+        context
+            .env
+            .add_diag(diag!(Syntax::InvalidModifier, (*l2, msg)));
+    } else if !postfix_abilities.is_empty() {
+        *prefix_abilities = postfix_abilities;
+    }
+    Ok(())
+}
+
 // Parse a struct variant. The returned boolean indicates whether the variant has a braced (`{..}`)
```

Comment:
Update the comment to indicate that the variant might have positional fields.

Reply:
done